contracts/Witch.sol

```diff
@@ -13,198 +13,358 @@ import "@yield-protocol/utils-v2/contracts/math/WDivUp.sol";
     event Auctioned(bytes12 indexed vaultId, uint256 indexed start);
+    event Bought(bytes12 indexed vaultId, address indexed buyer, uint256 ink, uint256 art);
+    event LineSet(bytes6 indexed ilkId, bytes6 baseId, uint32 duration, uint64 proportion, uint64 initialOffer);
+    event LimitSet(bytes6 indexed ilkId, bytes6 baseId, uint96 max, uint24 dust, uint8 dec);
+    event Point(bytes32 indexed param, address indexed value);
 
     struct Auction {
         address owner;
         uint32 start;
+        bytes6 baseId; // We cache the baseId here
+        uint128 art;
+        uint128 ink;
     }
 
-    struct Ilk {
-        uint32 duration;      // Time that auctions take to go to minimal price and stay there.
-        uint64 initialOffer;  // Proportion of collateral that is sold at auction start (1e18 = 100%)
+    struct Line {
+        uint32 duration; // Time that auctions take to go to minimal price and stay there
+        uint64 proportion; // Proportion of the vault that is available each auction (1e18 = 100%)
+        uint64 initialOffer; // Proportion of collateral that is sold at auction start (1e18 = 100%)
     }
 
     struct Limits {
-        uint96 line;                                                    // Maximum concurrent auctioned collateral
-        uint24 dust;                                                    // Minimum collateral that must be left when buying, unless buying all
-        uint8 dec;                                                      // Multiplying factor (10**dec) for line and dust
-        uint128 sum;                                                    // Current concurrent auctioned collateral
+        uint96 max; // Maximum concurrent auctioned collateral
+        uint24 dust; // Minimum collateral that must be left when buying, unless buying all
+        uint8 dec; // Multiplying factor (10**dec) for max and dust
+        uint128 sum; // Current concurrent auctioned collateral
     }
 
-    ICauldron immutable public cauldron;
+    ICauldron public immutable cauldron;
     ILadle public ladle;
     mapping(bytes12 => Auction) public auctions;
-    mapping(bytes6 => Ilk) public ilks;
-    mapping(bytes6 => Limits) public limits;
+    mapping(bytes6 => mapping(bytes6 => Line)) public lines;
+    mapping(bytes6 => mapping(bytes6 => Limits)) public limits;
 
-    constructor (ICauldron cauldron_, ILadle ladle_) {
+    constructor(ICauldron cauldron_, ILadle ladle_) {
         cauldron = cauldron_;
         ladle = ladle_;
     }
 
-
     /// @dev Point to a different ladle
+    /// @param param Name of parameter to set (must be "ladle")
+    /// @param value Address of new ladle
     function point(bytes32 param, address value) external auth {
-        if (param == "ladle") ladle = ILadle(value);
-        else revert("Unrecognized parameter");
+        require(param == "ladle", "Unrecognized");
+        ladle = ILadle(value);
         emit Point(param, value);
     }
```

Comment:
any reason for a generic method that can only set the ladle instead of something like `setLadle(address)`?

Reply:
Because all across the protocol we use that same method for reorchestration.